janus-idp/operator
# Security mitigation: remove secret get from RBAC

## Description
Security migration based on https://issues.redhat.com/browse/RHIDP-1012.
1)  Deploy the operator in a separate namespace:  add 'operatorframework.io/suggested-namespace=backstage-system' annotation to the CSV to indicate that the user is suggested to deploy the operator in a dedicated namespace such as backstage-system. For downstream, it will be named like openshift-rhdh-operator.

2) Remove the secret get/list permissions from RBAC.

a) Check and create the local db secret: https://github.com/janus-idp/operator/blob/main/controllers/local_db_secret.go#L54 – this can be re-implemented without the secret get call.
b) Read the secret data to get the list of files to mount:
[https://github.com/janus-idp/operator/blob/main/controllers/backstage_extra_files.go#L130:](https://github.com/janus-idp/operator/blob/main/controllers/backstage_extra_files.go#L130)  for this one, we will require the user to provide the keys and an error message will be logged if the keys are missing. This behavior shall be documented accordingly.
3) For custom image: we need to document that the user is responsible for the security compliance of the custom image.

- Fixes #153 

## PR acceptance criteria


## How to test changes / Special notes to the reviewer
1) Deploy the operator on OpenShift.
2) oc apply -f examples/janus-cr-with-app-configs.yaml
3) Check that backstage app is deployed and running:
```
❯ oc get pod
NAME                                                          READY   STATUS    RESTARTS   AGE
backstage-controller-manager-7f459bc647-75gz9                 2/2     Running   0          22m
backstage-my-backstage-app-with-app-config-6896c78f59-h5pc9   1/1     Running   0          9m47s
backstage-my-backstage-app-with-app-config-6896c78f59-jcgdw   1/1     Running   0          11m
backstage-psql-my-backstage-app-with-app-config-0             1/1     Running   0          21m
```
Check status of the CR:
```
status:
  conditions:
  - lastTransitionTime: "2024-01-29T15:45:53Z"
    message: ""
    reason: DeployOK
    status: "True"
    type: Deployed
``` 
4) Manually modify file examples/janus-cr-with-app-configs.yaml and remove the line "key: secret_file1.txt" and then save the file.
5) Update the CR:
oc apply -f examples/janus-cr-with-app-configs.yaml
6) Now you will see the error status:
```
status:
  conditions:
  - lastTransitionTime: "2024-01-29T15:53:12Z"
    message: 'failed to deploy Backstage Deployment: failed to add volume mounts to
      Backstage deployment, reason: key is required to mount extra file with secret
      my-backstage-extra-files-secret1'
    reason: DeployFailed
    status: "False"
```
